algorithms/kernel/service_hash_table.h

```diff
@@ -0,0 +1,189 @@
+};
+
+// Robert Jenkins' 32 bit integer hash function
+template <CpuType cpu>
+struct Hash<cpu, uint32_t>
+{
+    size_t operator()(uint32_t a) const
+    {
+        a = (a + 0x7ed55d16) + (a << 12);
+        a = (a ^ 0xc761c23c) ^ (a >> 19);
+        a = (a + 0x165667b1) + (a << 5);
+        a = (a + 0xd3a2646c) ^ (a << 9);
+        a = (a + 0xfd7046c5) + (a << 3);
+        a = (a ^ 0xb55a4f09) ^ (a >> 16);
+        return static_cast<size_t>(a);
+    }
+};
+
+template <CpuType cpu, typename KeyType, typename ValueType, typename HashFuncType = Hash<cpu, KeyType> >
+class HashTable
+{
+public:
+    HashTable(const size_t size) : _size(size), _table(size)
+    {
+        for (size_t i = 0; i < _size; ++i)
+        {
+            _table[i] = nullptr;
+        }
+    }
+
+    ~HashTable()
+    {
+        for (size_t i = 0; i < _size; ++i)
+        {
+            Entry * curr = _table[i];
+
+            while (curr)
+            {
+                Entry * next = curr->next;
+                delete curr;
+                curr = next;
+            }
+            _table[i] = nullptr;
+        }
+    }
+
+    bool find(const KeyType & key, ValueType & value)
+    {
+        const size_t id = _hashFunc(key) % _size;
+        Entry * entry   = _table[id];
+        while (entry)
+        {
+            if (entry->key == key)
+            {
+                value = entry->value;
+                return true;
+            }
+            entry = entry->next;
+        }
+        return false;
+    }
+
+    void insert(const KeyType & key, const ValueType & value)
+    {
+        const size_t id = _hashFunc(key) % _size;
+        Entry * entry   = _table[id];
+        Entry * prev    = nullptr;
+        while (entry && entry->key != key)
+        {
+            prev  = entry;
+            entry = entry->next;
+        }
+        if (entry)
+        {
+            entry->value = value;
+        }
+        else
+        {
+            entry = Entry::create(key, value);
+            if (!entry) return;
+            if (prev)
+            {
+                prev->next = entry;
+            }
+            else
+            {
+                _table[id] = entry;
+            }
+        }
+    }
+
+    void erase(const KeyType & key)
+    {
+        const size_t id = _hashFunc(key) % _size;
+        Entry * entry   = _table[id];
+        Entry * prev    = nullptr;
+
+        while (entry && entry->key != key)
+        {
+            prev  = entry;
+            entry = entry->next;
+        }
+        if (entry)
+        {
+            if (prev)
+            {
+                prev->next = entry->next;
+            }
+            else
+            {
+                _table[id] = entry->next;
+            }
+            delete entry;
+        }
+    }
+
+private:
+    struct Entry
+    {
+        DAAL_NEW_DELETE();
+
+        static Entry * create(const KeyType & key, const ValueType & value)
+        {
+            auto val = new Entry(key, value);
```

Comment:
So, you allocate many-many Entries with ```new```. In general allocation with ```new``` has some const memory consumption overhead for each allocation + overheads for memory alignment.

Are you sure, that we fit in required cache size in this case?

Reply:
Thank! I will try to reduce the number of allocation calls.

Reply:
Applied in this commit https://github.com/oneapi-src/oneDAL/pull/648/commits/d2c002dc7378c1008ada2cc63d321c2fb73e799b